lib/api/units/validateHeaders.js

```diff
@@ -0,0 +1,68 @@
+const { HeadersJsonExample } = require('../../validators/headers-json-example');
+
+const APIARY_JSON_HEADER_TYPE = 'application/vnd.apiary.http-headers+json';
+
+function normalizeHeaders(headers) {
+  return (
+    headers instanceof Object &&
+    Object.keys(headers).reduce(
+      (acc, headerName) =>
+        Object.assign({}, acc, {
+          [headerName.toLowerCase()]: headers[headerName]
+        }),
+      {}
+    )
+  );
+}
+
+function getHeadersType(headers) {
+  return headers instanceof Object && !Array.isArray(headers)
+    ? APIARY_JSON_HEADER_TYPE
+    : null;
+}
+
+/**
+ * Validates given real and expected headers.
+ * @param {Object} realHeaders
+ * @param {Object} expectedHeaders
+ */
+function validateHeaders(realHeaders, expectedHeaders) {
+  const real = normalizeHeaders(realHeaders);
+  const expected = normalizeHeaders(expectedHeaders);
+  const realType = getHeadersType(real);
+  const expectedType = getHeadersType(expected);
```

Comment:
The logic is simple, the code looks verbose. We're getting two objects and we want to validate them. We know their types (two objects) and we know we want to validate them as headers. It's either an error or `application/vnd.apiary.http-headers+json`. I didn't read the code, but what else could Gavel possibly send to this function?

Perhaps we could make it much simpler with all the assumptions:

```js
try {
  const real = normalizeHeaders(realHeaders);
  const expected = normalizeHeaders(expectedHeaders);
} catch (e) {
  return ... // got some rubbish
}

const validator = new HeadersJsonExample(real, expected);
const rawData = validator.validate();

...

return {
  validator: 'HeadersJsonExample',
  realType: 'application/vnd.apiary.http-headers+json',
  expectedType: 'application/vnd.apiary.http-headers+json',
  rawData,
  results
};
```

Reply:
But all the normalization/etc. functions are not expected to throw. The proper validator is assigned based on whether both header types are JSON. I'm not sure what try/catch would catch in this case.

Reply:
In that case it can be even simpler :) If we ensure the realHeaders and expectedHeaders are always {} on the input, then it's okay without any error handling

Reply:
Good idea, but I think that there is a feature that asserts that error would be thrown. We can rework this together with the spec in the further changes.